sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala

```diff
@@ -345,14 +345,23 @@ private[sql] class ParquetRelation2(
     // Schema of the whole table, including partition columns.
     var schema: StructType = _
 
+    // Cached leaves
+    val cachedLeaves: mutable.Map[Path, FileStatus] = new mutable.HashMap[Path, FileStatus]()
+
     /**
      * Refreshes `FileStatus`es, footers, partition spec, and table schema.
      */
     def refresh(): Unit = {
       // Lists `FileStatus`es of all leaf nodes (files) under all base directories.
+      // We only care about new FileStatus and updated FileStatus.
       val leaves = cachedLeafStatuses().filter { f =>
-        isSummaryFile(f.getPath) ||
-          !(f.getPath.getName.startsWith("_") || f.getPath.getName.startsWith("."))
+        (isSummaryFile(f.getPath) ||
+          !(f.getPath.getName.startsWith("_") || f.getPath.getName.startsWith("."))) &&
+          (!cachedLeaves.contains(f.getPath) ||
+            cachedLeaves(f.getPath).getModificationTime < f.getModificationTime)
+      }.map { f =>
+        cachedLeaves += (f.getPath -> f)
+        f
```

Comment:
I guess you are trying to only keep `FileStatus`es of those files that need to be touched during schema merging here. Actually the `FileStatus` cache must be consistent with the files stored on the file system. Because we also inject the cache to `ParquetInputFormat` in `ParquetRelation2.buildScan` to avoid calling `listStatus` repeatedly there.